sklearn/utils/estimator_checks.py

```diff
@@ -395,57 +402,145 @@ def _yield_all_checks(estimator, legacy: bool):
-    # This is similar to _maybe_mark_xfail(), but this one is used by
-    # check_estimator() instead of @parametrize_with_checks which requires
-    # pytest
-    should_be_skipped, reason = _should_be_skipped_or_marked(estimator, check)
-    if not should_be_skipped:
-        return check
+        return estimator, wrapped
 
-    check_name = check.func.__name__ if isinstance(check, partial) else check.__name__
 
-    @wraps(check)
-    def wrapped(*args, **kwargs):
-        raise SkipTest(
-            f"Skipping {check_name} for {estimator.__class__.__name__}: {reason}"
-        )
+def _should_be_skipped_or_marked(
+    estimator, check, expected_failed_checks: dict[str, str] | None = None
+) -> tuple[bool, str]:
+    """Check whether a check should be skipped or marked as xfail.
+
+    Parameters
+    ----------
+    estimator : estimator object
+        Estimator instance for which to generate checks.
+    check : partial or callable
+        Check to be marked.
+    expected_failed_checks : dict[str, str], default=None
+        Dictionary of the form {check_name: reason} for checks that are expected to
+        fail.
 
-    return wrapped
+    Returns
+    -------
+    should_be_marked : bool
+        Whether the check should be marked as xfail or skipped.
+    reason : str
+        Reason for skipping the check.
+    """
+
+    expected_failed_checks = expected_failed_checks or {}
+
+    check_name = _check_name(check)
+    if check_name in expected_failed_checks:
+        return True, expected_failed_checks[check_name]
 
+    return False, "Check is not expected to fail"
 
-def _should_be_skipped_or_marked(estimator, check):
-    # Return whether a check should be skipped (when using check_estimator())
-    # or marked as XFAIL (when using @parametrize_with_checks()), along with a
-    # reason.
-    # Currently, a check should be skipped or marked if
-    # the check is in the _xfail_checks tag of the estimator
 
-    check_name = check.func.__name__ if isinstance(check, partial) else check.__name__
+def checks_generator(
+    estimator,
+    *,
+    legacy: bool = True,
+    expected_failed_checks: dict[str, str] | None = None,
+    mark: Literal["xfail", "skip", None] = None,
+):
+    """Generate checks for an estimator.
+
+    Parameters
+    ----------
+    estimator : estimator object
+        Estimator instance for which to generate checks.
+    legacy : bool, default=True
+        Whether to include legacy checks. Over time we remove checks from this category
+        and move them into their specific category.
+    expected_failed_checks : dict[str, str], default=None
+        Dictionary of the form {check_name: reason} for checks that are expected to
+        fail.
+    mark : "xfail" or "skip" or None, default=None
+        Whether to mark the checks that are expected to fail as
+        xfail(`pytest.mark.xfail`) or skip. Skip will raise a
+        :class:`~sklearn.exceptions.SkipTest` for the test.
```

Comment:
It's not clear as to when this exception will be raised. I think this should be rephrased to be more explicit.

Reply:
I think it's clear now.

Reply:
It is, thanks.